File: 05-composicion-estaticos/python/alumno.py
"""
alumno.py

REGLA DE ORO PARA DISTINGUIR, mirando este archivo:
  ¿va en __init__ con self.?  -> es de cada objeto
  ¿va suelto dentro de class? -> es de la clase, compartido por todos

  static int totalCreados = 0;   ->  total_creados = 0   (DENTRO de la clase,
                                                          FUERA de __init__)
  static boolean aprueba(...)    ->  @staticmethod
  static int getTotalCreados()   ->  @classmethod  (recibe cls = la clase)
  static final String COLEGIO    ->  COLEGIO = "..."  (Python no tiene constantes
                                                       reales: MAYUSCULAS = "no tocar")
"""


class Alumno:

    # ---- ATRIBUTOS DE CLASE (el "static" de Java) ----
    total_creados = 0
    COLEGIO = "Colegio Central"
    NOTA_MINIMA = 4.0

    def __init__(self, nombre, nota):
        # ---- ATRIBUTOS DE INSTANCIA: uno por objeto ----
        self._nombre = nombre
        self._nota = nota
        # Ojo: se incrementa en la CLASE, no en self
        Alumno.total_creados += 1

    def get_nombre(self):
        return self._nombre

    def get_nota(self):
        return self._nota

    @staticmethod
    def aprueba(nota):
        """No usa self ni cls: es una funcion util que vive dentro de la clase."""
        return nota >= Alumno.NOTA_MINIMA

    @classmethod
    def get_total_creados(cls):
        """Recibe la CLASE (cls), no el objeto."""
        return cls.total_creados

File: 05-composicion-estaticos/java/Alumno.java
/*
 * Alumno.java
 *
 * Ojo a la mezcla que hay en este archivo:
 *
 *   static  -> pertenece a la CLASE. Existe UNA sola copia para todo el
 *              programa, aunque crees mil alumnos.
 *   normal  -> pertenece a cada OBJETO. Cada alumno tiene el suyo.
 *
 * Regla practica: si al leerlo piensas "esto es de Camila", va sin static.
 * Si piensas "esto es de los alumnos en general", va con static.
 */
public class Alumno {

    // STATIC = pertenece a la CLASE. Una sola copia para todo el programa.
    private static int totalCreados = 0;

    // STATIC FINAL = constante de clase (se escribe en MAYUSCULAS por convencion)
    public static final String COLEGIO = "Colegio Central";
    public static final double NOTA_MINIMA = 4.0;

    // Atributos de INSTANCIA: cada alumno tiene los suyos
    private String nombre;
    private double nota;

    public Alumno(String nombre, double nota) {
        this.nombre = nombre;
        this.nota = nota;
        totalCreados++;            // sin "this": no es de este objeto, es de la clase
    }

    public String getNombre() { return this.nombre; }
    public double getNota()   { return this.nota; }

    // Metodo estatico: no usa "this", no necesita un objeto para funcionar
    public static boolean aprueba(double nota) {
        return nota >= NOTA_MINIMA;
    }

    public static int getTotalCreados() {
        return totalCreados;
    }
}
